### PR TITLE
Added Fast Cells and GRU support to SRNN, Corrected Readme

### PR DESCRIPTION
**Pull request by SachinG007**

### CODE REVIEW

File: pytorch/pytorch_edgeml/graph/rnn.py [hunk 1]
```diff
@@ -1050,7 +1050,7 @@ def forward(self, input, hiddenState=None, cellState=None):
 
 class SRNN2(nn.Module):
 
-    def __init__(self, inputDim, outputDim, hiddenDim0, hiddenDim1, cellType):
+    def __init__(self, inputDim, outputDim, hiddenDim0, hiddenDim1, cellType, **fastCellArgs):
```
Comment (metastableB) [inline pytorch/pytorch_edgeml/graph/rnn.py]: This should be a generic `cellArgs` so for any cell that might be used to day or in the future. Also, make it a dictionary not optional arguments.

File: pytorch/pytorch_edgeml/graph/rnn.py [hunk 2]
```diff
@@ -1065,14 +1065,33 @@ def __init__(self, inputDim, outputDim, hiddenDim0, hiddenDim1, cellType):
         self.hiddenDim0 = hiddenDim0
         self.hiddenDim1 = hiddenDim1
         self.outputDim = outputDim
-        supportedCells = ['LSTM']
+        supportedCells = ['LSTM','FastGRNNCell','GRULRCell']
```
Comment (metastableB) [inline pytorch/pytorch_edgeml/graph/rnn.py]: What about FastRNN ?

File: pytorch/pytorch_edgeml/graph/rnn.py [hunk 3]
```diff
@@ -1065,14 +1065,33 @@ def __init__(self, inputDim, outputDim, hiddenDim0, hiddenDim1, cellType):
...
             self.rnnClass = nn.LSTM
-
-        self.rnn0 = self.rnnClass(input_size=inputDim, hidden_size=hiddenDim0)
-        self.rnn1 = self.rnnClass(input_size=hiddenDim0, hidden_size=hiddenDim1)
+            self.rnn0 = self.rnnClass(input_size=inputDim, hidden_size=hiddenDim0)
+            self.rnn1 = self.rnnClass(input_size=hiddenDim0, hidden_size=hiddenDim1)
+        
+        if self.cellType == 'FastGRNNCell':
+            self.rnnClass = FastGRNN
+            self.rnn0 = self.rnnClass(input_size=inputDim, hidden_size=hiddenDim0, gate_non_linearity = fastCellArgs['gate_non_linearity'],
```
Comment (metastableB) [inline pytorch/pytorch_edgeml/graph/rnn.py]: Please use PEP8 to break lines properly and format properly.

Comment (metastableB) [reply pytorch/pytorch_edgeml/graph/rnn.py]: More importantly, this is bad code. What if the user did not provided `gate_non_linearity` and instead want to use the default arguments in FastCells. This step will fail with a `KeyNotFound` error.

File: pytorch/pytorch_edgeml/graph/rnn.py [hunk 4]
```diff
@@ -1065,14 +1065,33 @@ def __init__(self, inputDim, outputDim, hiddenDim0, hiddenDim1, cellType):
...
+                 update_non_linearity = fastCellArgs['update_non_linearity'], wRank = fastCellArgs['wRank'], uRank = fastCellArgs['uRank'],
```
Comment (metastableB) [inline pytorch/pytorch_edgeml/graph/rnn.py]: Please do not extend more than 80 chars. This is difficult to review.

File: pytorch/pytorch_edgeml/graph/rnn.py [hunk 5]
```diff
@@ -1065,14 +1065,33 @@ def __init__(self, inputDim, outputDim, hiddenDim0, hiddenDim1, cellType):
...
+                 zetaInit = fastCellArgs['zetaInit'], nuInit = fastCellArgs['nuInit'], batch_first = fastCellArgs['batch_first'])
+            self.rnn1 = self.rnnClass(input_size=hiddenDim0, hidden_size=hiddenDim1, gate_non_linearity = fastCellArgs['gate_non_linearity'],
+                 update_non_linearity = fastCellArgs['update_non_linearity'], wRank = fastCellArgs['wRank'], uRank = fastCellArgs['uRank'],
+                 zetaInit = fastCellArgs['zetaInit'], nuInit = fastCellArgs['nuInit'], batch_first = fastCellArgs['batch_first'])
```
Comment (metastableB) [inline pytorch/pytorch_edgeml/graph/rnn.py]: You don't need any of these steps. Once you have made `cellArgs` a dictionary argument of the `__init__` function, you just have to do:

```
self.rnn0 = self.rnnClass(input_size=inputDim, hidden_size=hiddenDim0, **cellArgs)
```

File: pytorch/pytorch_edgeml/graph/rnn.py [hunk 6]
```diff
@@ -1065,14 +1065,33 @@ def __init__(self, inputDim, outputDim, hiddenDim0, hiddenDim1, cellType):
...
+
+        if self.cellType == 'GRULRCell':
+            self.rnnClass = GRU
+            self.rnn0 = self.rnnClass(input_size=inputDim, hidden_size=hiddenDim0, gate_non_linearity = fastCellArgs['gate_non_linearity'],
+                 update_non_linearity = fastCellArgs['update_non_linearity'], wRank = fastCellArgs['wRank'], uRank = fastCellArgs['uRank'])
```
Comment (metastableB) [inline pytorch/pytorch_edgeml/graph/rnn.py]: What is the use of extracting these arguments from the dictionary here?

File: pytorch/pytorch_edgeml/graph/rnn.py [hunk 7]
```diff
@@ -1117,12 +1136,12 @@ def forward(self, x, brickSize):
         x_bricks = torch.reshape(x_bricks, [oldShape[0], oldShape[1] * oldShape[2], oldShape[3]])
         # x bricks: [brickSize, numBricks * batchSize, featureDim]
         # x_bricks = torch.Tensor(x_bricks)
-        hidd0, out0 = self.rnn0(x_bricks)
+        hidd0 = self.rnn0(x_bricks)
```
Comment (metastableB) [inline pytorch/pytorch_edgeml/graph/rnn.py]: what happened to output ? Earlier this was 
```
hidd0, out0 = self.rnn0(x_bricks)
```

**Comment by metastableB** [comment]
@SachinG007 

The code should follow the following strucutre:

```
if cellType == 'A':
    rnnClass = A
elif cellType == 'B':
    rnnClass = B
else:
    rnnClass = C

rnn0 = rnnClass(inp, out, **cellArgs)
rnn1 = rnnClass(inp, out, **cellArgs)
```

File: pytorch/pytorch_edgeml/graph/rnn.py (hunk 5 above)
Comment (SachinG007) [reply pytorch/pytorch_edgeml/graph/rnn.py]: Resolved

File: pytorch/pytorch_edgeml/graph/rnn.py (hunk 6 above)
Comment (SachinG007) [reply pytorch/pytorch_edgeml/graph/rnn.py]: My Bad

File: pytorch/pytorch_edgeml/graph/rnn.py (hunk 2 above)
Comment (SachinG007) [reply pytorch/pytorch_edgeml/graph/rnn.py]: Done

File: pytorch/pytorch_edgeml/graph/rnn.py (hunk 1 above)
Comment (SachinG007) [reply pytorch/pytorch_edgeml/graph/rnn.py]: Ok

File: pytorch/pytorch_edgeml/graph/rnn.py (hunk 7 above)
Comment (SachinG007) [reply pytorch/pytorch_edgeml/graph/rnn.py]: @adityakusupati  I was pointing to this in the mail

**Pull request closed without merging**